training/controllers/auth.py

```diff
@@ -0,0 +1,92 @@
+from flask import Blueprint, request, redirect, render_template, url_for, g, session
+
+from training.extensions import db, bcrypt
+# Duda, porque no funciona?
+# with current_app.app_context():
+#    db = current_app.db
+from training.models.users import User
+from training.controllers.forms import index_form, login_form
+from training.controllers.function_decorators import login_required
+
+
+bp = Blueprint('name1', __name__)
+
+
+@bp.before_app_request
+def user_to_g():
+    username = session.get('username')
+    if username is not None:
+        # If a user has a session, it will be always in the database
+        user = db.session.query(User).filter_by(id=username).first()
+        g.user = user
+    else:
+        g.user = None
+    return None
+
+
+def test_unique_fields(form):
+    if User.query.filter_by(id=form['username']).first() or User.query.filter_by(email=form['email']).first() is not None:
+        return False
+    else:
+        return True
```

Comment:
Can't we add these validations to the model?

Reply:
The point is that what we verify here, is that when a new user submits a registration form, he could be trying to register with a valid email, but already registered.

So here we query the database to check if a new user is trying to register with an already registered email, even if the email form was validated.

Reply:
I understand. But the uniqueness of email and username is a validation too. Maybe SQLAlchemy or WTForms have a feature to do this so you just call form.validate().